src/util/small_shared_n_way_ptr.h

```diff
@@ -0,0 +1,389 @@
+    }
+
+    destructt<num_types - 1, pointee_baset, Ts...>::destruct(p);
+  }
+
+  pointee_baset *p = nullptr;
+};
+
+/// Constructs a small shared n-way pointer, with two possible pointee types
+/// (i.e., n = 2), and constructs an object of either type U (when I = 0) or
+/// type V (when I = 1) pointed to by the shared pointer. Arguments ts are
+/// passed to the constructor of U or V. U and V must be subclasses of
+/// small_shared_n_way_pointee_baset<2, Num>.
+///
+/// \tparam I index of the type of object to construct (0 -> U, 1 -> V)
+/// \tparam U first possible pointee type
+/// \tparam V second possible pointee type
+/// \tparam Ts types of arguments to pass to the constructor of U or V
+/// \param ts arguments to pass to the constructor of U or V
+template <std::size_t I, typename U, typename V, typename... Ts>
+small_shared_n_way_ptrt<U, V> make_shared_2(Ts &&... ts)
+{
+  return small_shared_n_way_ptrt<U, V>::template create_small_shared_n_way_ptr<
+    I>(new typename get_typet<I, U, V>::type(std::forward<Ts>(ts)...));
+}
+
+/// Constructs a small shared n-way pointer, with three possible pointee types
+/// (i.e., n = 3), and constructs an object of either type U (when I = 0), type
+/// V (when I = 1), or type W (when I = 2), pointed to by the shared pointer.
+/// Arguments ts are passed to the constructor of U, V, or W. U, V, and W must
+/// be subclasses of small_shared_n_way_pointee_baset<3, Num>.
+///
+/// \tparam I index of the type of object to construct (0 -> U, 1 -> V)
+/// \tparam U first possible pointee type
+/// \tparam V second possible pointee type
+/// \tparam W third possible pointee type
+/// \tparam Ts types of arguments to pass to the constructor of U or V
+/// \param ts arguments to pass to the constructor of U or V
+template <std::size_t I, typename U, typename V, typename W, typename... Ts>
+small_shared_n_way_ptrt<U, V, W> make_shared_3(Ts &&... ts)
+{
+  return small_shared_n_way_ptrt<U, V, W>::
+    template create_small_shared_n_way_ptr<I>(
+      new typename get_typet<I, U, V, W>::type(std::forward<Ts>(ts)...));
+}
+
+template <typename... Ts>
+bool operator==(
+  const small_shared_n_way_ptrt<Ts...> &lhs,
+  const small_shared_n_way_ptrt<Ts...> &rhs)
+{
+  return lhs.get() == rhs.get();
+}
+
+template <typename... Ts>
+bool operator!=(
+  const small_shared_n_way_ptrt<Ts...> &lhs,
+  const small_shared_n_way_ptrt<Ts...> &rhs)
+{
+  return lhs.get() != rhs.get();
+}
+
+template <std::size_t N, typename Num>
+class small_shared_n_way_pointee_baset
+{
+public:
+  static_assert(std::is_unsigned<Num>::value, "Num must be an unsigned type");
+
+  small_shared_n_way_pointee_baset() = default;
+
+  // The use count shall be unaffected
+  small_shared_n_way_pointee_baset(const small_shared_n_way_pointee_baset &)
+  {
+  }
+
+  // The use count shall be unaffected
+  small_shared_n_way_pointee_baset &
+  operator=(const small_shared_n_way_pointee_baset &)
+  {
+    return *this;
+  }
+
+  Num use_count() const
```

Comment:
Could we actually name this `get_use_count` to avoid the final underscore in `use_count_`? It would also seem a lot more consistent, because we have `increment_use_count` and `decrement_use_count`.